### ISSUE TITLE
JDBC session exception handling is broken

### ISSUE THREAD

**Issue by piotrkot** [jcabi/jcabi-jdbc#100]
Consider the code:

```
new JdbcSession(source)
      .autocommit(false)
      .sql("START TRANSACTION")
      .execute()
      .sql("DELETE FROM employee WHERE name = ?")
      .set("Jeff Lebowski")
      .execute()
      .sql("INSERT INTO employee VALUES (?)")
      .set(this.username())
      .execute()
      .commit();

private String username() throws IOException {
    throw new IOException();
}
```

When run as a unit test is hangs because transaction is not rollbacked.
`JdbcSession.run()` method should rollback on any exception and not just on the `SQLException` type.

**Comment by piotrkot** [comment]
@yegor256 Not waiting for the fix, I wanted to create a PR. But the advised in README command `mvn clean install -Pqulice` does not execute integration tests. The test in this issue must be run as an integration one.
Can you help running integration tests in jcabi-jdbc?

**Comment by yegor256** [comment]
@piotrkot I think it does execute integration tests. Did you try?

**Comment by piotrkot** [comment]
@yegor256 I wouldn't bother you otherwise :)

When I added
```
    /**
     * Fails.
     * @throws Exception If there is some problem inside
     */
    @Test
    void shouldFail() throws Exception {
        MatcherAssert.assertThat(false, Matchers.is(true));
    }
```
to the `JdbcSessionITCase` build is successful.

**Comment by piotrkot** [comment]
I can find this in the build logs:
```
[INFO] --- maven-failsafe-plugin:3.0.0-M5:integration-test (default) @ jcabi-jdbc --- 
[WARNING] useSystemClassLoader setting has no effect when not forking                          
[WARNING] The parameter forkCount should likely not be 0, not forking a JVM for tests reduce test accuracy, ensure to have a <forkCount> >= 1.                                                                     
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".                                                                                                                                      
SLF4J: Defaulting to no-operation (NOP) logger implementation                       
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.                     
[INFO]                                                                                                
[INFO] Results:                                                                                     
[INFO]                                                                                                                                                                                                             
[INFO] Tests run: 0, Failures: 0, Errors: 0, Skipped: 0                                  
[INFO]                                                                                 
[INFO]                                                                             
[INFO] --- maven-failsafe-plugin:3.0.0-M5:verify (default) @ jcabi-jdbc ---  
```
which doesn't look good. Do you know why is that?

**Comment by piotrkot** [comment]
Let's make it a separate issue - #101.

**Comment by andreoss** [comment]
@piotrkot 
The rollback only make sense for `SqlException` which are caught inside `JdbcSession`. 
In your example the exception is thrown at line 2, before `execute`.
```
 1)   .sql("INSERT INTO employee VALUES (?)")
 2)   .set(this.username()) /// Thrown at this line
 3)   .execute()
```

**Comment by piotrkot** [comment]
@andreoss When you provide arguments to the `JdbcSession` from other methods, it may happen they fail with exceptions other than `SqlException`. If this is not caught, the connection remains open. I believe connection should be closed as soon as they are not needed (it also applies when using connection pools).

**Comment by andreoss** [comment]
@piotrkot In code your provided IOException will happend before method `.set` or `.executre` are called.  

This is how the object you pass with `set` is handled:
https://github.com/jcabi/jcabi-jdbc/blob/master/src/main/java/com/jcabi/jdbc/PrepareArgs.java#L68

**Comment by piotrkot** [comment]
@andreoss How about this code?
```
JdbcSession session = new JdbcSession(source)
      .autocommit(false)
      .sql("START TRANSACTION")
      .execute()
      .sql("DELETE FROM employee WHERE name = ?")
      .set("Jeff Lebowski")
      .execute();
session.sql("INSERT INTO employee VALUES (?)")
      .set(this.username())
      .execute()
      .commit();

private String username() throws IOException {
    throw new IOException();
}
```

**Comment by andreoss** [comment]
@piotrkot So you mean that any incomplete transaction should be rollbacked?
 I guess in this case you should handle this situation yourself and call `session.rollback` explicitly inside a catch block for example.
A possible solution could be also having a special method for doing tansactions instead of "START TRANSACTION", i.e
```
new JdbcSession(source)
    .transaction(session -> { ..... })
    .commit();
```

**Comment by piotrkot** [comment]
@andreoss Sure, you could have a method `transaction()`. I took the example right from the [documentation](https://jdbc.jcabi.com/example-transaction.html).
Very likely that is the problem of the builder pattern used.

As a matter of fact, right after filing this bug I created a fork of this library - https://github.com/piotrkot/oojdbc - and my problem is fixed. But with this bug I wanted to let the maintainers know there is something to improve here.

**Comment by andreoss** [comment]
@piotrkot Closed in favor of #106

**Comment by 0crat** [comment]
Job `gh:jcabi/jcabi-jdbc#100` is not assigned, can't get performer
